File: src/routes/subscriptions.py
"""Subscription routes: /api/subscriptions"""
from flask import Blueprint, request, jsonify, g
from src.db.init_db import get_db
from src.middleware.auth import require_auth, require_permission

subscriptions_bp = Blueprint("subscriptions", __name__, url_prefix="/api/subscriptions")


@subscriptions_bp.get("/")
@require_auth
@require_permission("VIEW_SUBSCRIPTIONS")
def list_subscriptions():
    db = get_db()
    page = max(1, request.args.get("page", 1, type=int))
    per_page = min(100, request.args.get("per_page", 20, type=int))
    status = request.args.get("status")  # "active" | "cancelled" | "paused"
    offset = (page - 1) * per_page

    where_parts = []
    params = []

    if status == "active":
        where_parts.append("s.renewal_date >= DATE('now')")
    elif status == "cancelled":
        where_parts.append("s.renewal_date IS NULL AND s.auto_renew = 0")
    elif status == "paused":
        where_parts.append("s.auto_renew = 0 AND s.renewal_date IS NOT NULL")

    where = ("WHERE " + " AND ".join(where_parts)) if where_parts else ""

    base_q = f"""
        FROM subscription s
        JOIN subscription_plan sp ON s.plan_id = sp.plan_id
        JOIN customer c ON s.cus_id = c.customer_id
        {where}
    """
    total = db.execute(f"SELECT COUNT(*) {base_q}", params).fetchone()[0]
    rows = db.execute(f"""
        SELECT s.*, sp.plan_name, sp.price, sp.duration_months,
               c.first_name, c.last_name, c.email
        {base_q}
        ORDER BY s.start_date DESC LIMIT ? OFFSET ?
    """, params + [per_page, offset]).fetchall()
    db.close()

    return jsonify({
        "data": [dict(r) for r in rows],
        "pagination": {"page": page, "per_page": per_page, "total": total,
                       "pages": (total + per_page - 1) // per_page}
    })


@subscriptions_bp.get("/<int:subscription_id>")
@require_auth
@require_permission("VIEW_SUBSCRIPTIONS")
def get_subscription(subscription_id):
    db = get_db()
    sub = db.execute("""
        SELECT s.*, sp.plan_name, sp.price, sp.duration_months, sp.is_prepaid,
               c.first_name, c.last_name, c.email
        FROM subscription s
        JOIN subscription_plan sp ON s.plan_id = sp.plan_id
        JOIN customer c ON s.cus_id = c.customer_id
        WHERE s.subscription_id = ?
    """, (subscription_id,)).fetchone()

    if not sub:
        db.close()
        return jsonify({"error": "Subscription not found"}), 404

    events = db.execute(
        "SELECT * FROM subscription_event WHERE subscription_id = ? ORDER BY event_date DESC",
        (subscription_id,)
    ).fetchall()

    orders = db.execute("""
        SELECT co.*, br.release_month FROM customer_order co
        JOIN box_release br ON co.release_id = br.release_id
        WHERE co.subscription_id = ? ORDER BY co.created_at DESC
    """, (subscription_id,)).fetchall()

    db.close()
    return jsonify({
        **dict(sub),
        "events": [dict(e) for e in events],
        "orders": [dict(o) for o in orders],
    })


@subscriptions_bp.post("/")
@require_auth
@require_permission("EDIT_SUBSCRIPTIONS")
def create_subscription():
    data = request.get_json(silent=True) or {}
    required = ["plan_id", "product_id", "cus_id", "start_date"]
    missing = [f for f in required if not data.get(f)]
    if missing:
        return jsonify({"error": f"Missing required fields: {', '.join(missing)}"}), 400

    db = get_db()
    try:
        cur = db.execute("""
            INSERT INTO subscription (plan_id, product_id, cus_id, renewal_date, start_date, auto_renew)
            VALUES (?,?,?,?,?,?)
        """, (
            data["plan_id"], data["product_id"], data["cus_id"],
            data.get("renewal_date"), data["start_date"],
            int(data.get("auto_renew", 1))
        ))
        subscription_id = cur.lastrowid
        db.execute(
            "INSERT INTO subscription_event (subscription_id, event_type, event_date) VALUES (?,?,DATE('now'))",
            (subscription_id, "Created")
        )
        db.commit()
    except Exception as e:
        db.close()
        return jsonify({"error": str(e)}), 400

    db.close()
    return jsonify({"message": "Subscription created", "subscription_id": subscription_id}), 201


@subscriptions_bp.patch("/<int:subscription_id>/cancel")
@require_auth
@require_permission("EDIT_SUBSCRIPTIONS")
def cancel_subscription(subscription_id):
    db = get_db()
    sub = db.execute("SELECT 1 FROM subscription WHERE subscription_id = ?", (subscription_id,)).fetchone()
    if not sub:
        db.close()
        return jsonify({"error": "Subscription not found"}), 404

    db.execute(
        "UPDATE subscription SET auto_renew = 0, renewal_date = NULL WHERE subscription_id = ?",
        (subscription_id,)
    )
    db.execute(
        "INSERT INTO subscription_event (subscription_id, event_type, event_date) VALUES (?,?,DATE('now'))",
        (subscription_id, "Cancelled")
    )
    db.execute(
        "INSERT INTO audit_log (user_id,entity_name,entity_id,action_type,old_value,new_value) VALUES (?,?,?,?,?,?)",
        (g.user["user_id"], "subscription", subscription_id, "UPDATE", "status: Active", "status: Cancelled")
    )
    db.commit()
    db.close()
    return jsonify({"message": "Subscription cancelled"})


@subscriptions_bp.patch("/<int:subscription_id>/pause")
@require_auth
@require_permission("EDIT_SUBSCRIPTIONS")
def pause_subscription(subscription_id):
    db = get_db()
    sub = db.execute("SELECT 1 FROM subscription WHERE subscription_id = ?", (subscription_id,)).fetchone()
    if not sub:
        db.close()
        return jsonify({"error": "Subscription not found"}), 404

    db.execute(
        "UPDATE subscription SET auto_renew = 0 WHERE subscription_id = ?",
        (subscription_id,)
    )
    db.execute(
        "INSERT INTO subscription_event (subscription_id, event_type, event_date) VALUES (?,?,DATE('now'))",
        (subscription_id, "Paused")
    )
    db.commit()
    db.close()
    return jsonify({"message": "Subscription paused"})


@subscriptions_bp.get("/plans")
@require_auth
def list_plans():
    db = get_db()
    plans = db.execute("SELECT * FROM subscription_plan ORDER BY price").fetchall()
    db.close()
    return jsonify([dict(p) for p in plans])

File: functions/src/routes/subscriptions.ts
import { Router, Response } from "express";
import * as admin from "firebase-admin";
import { db } from "../index";
import { requireAuth, requirePermission, AuthRequest, writeAudit } from "../middleware/auth";

export const subscriptionsRouter = Router();

/** GET /api/subscriptions/plans — public */
subscriptionsRouter.get("/plans", async (_req, res: Response): Promise<void> => {
  const snap = await db.collection("subscription_plans").orderBy("price").get();
  res.json(snap.docs.map((d) => ({ id: d.id, ...d.data() })));
});

/** GET /api/subscriptions */
subscriptionsRouter.get(
  "/",
  requireAuth,
  requirePermission("VIEW_SUBSCRIPTIONS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const pageSize = Math.min(100, parseInt(req.query.per_page as string) || 20);
      const status   = req.query.status as string | undefined;

      let query: admin.firestore.Query = db.collection("subscriptions").orderBy("start_date", "desc");

      if (status === "active") {
        query = query.where("renewal_date", ">=", new Date().toISOString().split("T")[0])
                     .where("auto_renew", "==", true);
      } else if (status === "cancelled") {
        query = query.where("auto_renew", "==", false);
      }

      const afterDoc = req.query.after as string | undefined;
      if (afterDoc) {
        const snap = await db.collection("subscriptions").doc(afterDoc).get();
        if (snap.exists) query = query.startAfter(snap);
      }

      const snapshot = await query.limit(pageSize).get();
      res.json({
        data: snapshot.docs.map((d) => ({ id: d.id, ...d.data() })),
        next_cursor: snapshot.docs.length === pageSize
          ? snapshot.docs[snapshot.docs.length - 1].id : null,
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/subscriptions/:id */
subscriptionsRouter.get(
  "/:id",
  requireAuth,
  requirePermission("VIEW_SUBSCRIPTIONS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const doc = await db.collection("subscriptions").doc(req.params.id).get();
      if (!doc.exists) { res.status(404).json({ error: "Subscription not found" }); return; }

      const [events, orders] = await Promise.all([
        db.collection("subscriptions").doc(req.params.id)
          .collection("events").orderBy("event_date", "desc").limit(20).get(),
        db.collection("orders").where("subscription_id", "==", req.params.id)
          .orderBy("created_at", "desc").limit(20).get(),
      ]);

      res.json({
        id: doc.id,
        ...doc.data(),
        events: events.docs.map((d) => ({ id: d.id, ...d.data() })),
        orders: orders.docs.map((d) => ({ id: d.id, ...d.data() })),
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** POST /api/subscriptions — create */
subscriptionsRouter.post(
  "/",
  requireAuth,
  requirePermission("EDIT_SUBSCRIPTIONS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const { plan_id, product_id, customer_id, start_date } = req.body;
      if (!plan_id || !product_id || !customer_id || !start_date) {
        res.status(400).json({ error: "Missing required fields: plan_id, product_id, customer_id, start_date" });
        return;
      }

      const ref = await db.collection("subscriptions").add({
        plan_id, product_id, customer_id,
        renewal_date: req.body.renewal_date ?? null,
        start_date,
        auto_renew:  req.body.auto_renew ?? true,
        created_at:  admin.firestore.FieldValue.serverTimestamp(),
      });

      await db.collection("subscriptions").doc(ref.id)
        .collection("events").add({
          event_type: "Created",
          event_date: new Date().toISOString().split("T")[0],
        });

      res.status(201).json({ message: "Subscription created", subscription_id: ref.id });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** PATCH /api/subscriptions/:id/cancel */
subscriptionsRouter.patch(
  "/:id/cancel",
  requireAuth,
  requirePermission("EDIT_SUBSCRIPTIONS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const doc = await db.collection("subscriptions").doc(req.params.id).get();
      if (!doc.exists) { res.status(404).json({ error: "Not found" }); return; }

      await doc.ref.update({
        auto_renew: false, renewal_date: null,
        updated_at: admin.firestore.FieldValue.serverTimestamp(),
      });
      await doc.ref.collection("events").add({
        event_type: "Cancelled",
        event_date: new Date().toISOString().split("T")[0],
      });
      await writeAudit(req.user!.uid, "subscription", req.params.id, "UPDATE", "status: Active", "status: Cancelled");

      res.json({ message: "Subscription cancelled" });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** PATCH /api/subscriptions/:id/pause */
subscriptionsRouter.patch(
  "/:id/pause",
  requireAuth,
  requirePermission("EDIT_SUBSCRIPTIONS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const doc = await db.collection("subscriptions").doc(req.params.id).get();
      if (!doc.exists) { res.status(404).json({ error: "Not found" }); return; }

      await doc.ref.update({
        auto_renew: false,
        updated_at: admin.firestore.FieldValue.serverTimestamp(),
      });
      await doc.ref.collection("events").add({
        event_type: "Paused",
        event_date: new Date().toISOString().split("T")[0],
      });

      res.json({ message: "Subscription paused" });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);
